### ISSUE TITLE
panicked at 'called `Result::unwrap()` on an `Err` value: MaxCapacity(8096)'

### ISSUE THREAD

**Issue by jmpesp** [oxidecomputer/opte#346]
I was trying to reproduce the hang seen in https://github.com/oxidecomputer/propolis/issues/322 today on my setup, but managed to panic the host machine instead!

```
WARNING: panicked at 'called `Result::unwrap()` on an `Err` value: MaxCapacity(8096)', /work/oxidecomputer/opte/opte/src/engine/port.rs:1890:17

panic[cpu26]/thread=fffffeb1da1177a0: panicked at 'called `Result::unwrap()` on an `Err` value: MaxCapacity(8096)', /work/oxidecomputer/opte/opte/src/engine/port.rs:1890:17

fffffe00f732ee40 xde:rust_begin_unwind+146 ()
fffffe00f732ee80 xde:_ZN4core9panicking9panic_fmt17hbd42276bcaa2dfd8E+2d ()
fffffe00f732ef00 xde:_ZN4core6result13unwrap_failed17h5b67be868f6dc91bE+6c ()
fffffe00f732f310 xde:_ZN4opte6engine4port13Port$LT$N$GT$7process17h9bca109443cf9deaE+2bdf ()
fffffe00f732fc40 xde:xde_mc_tx+32f ()
fffffe00f732fc70 mac:mac_ring_tx+43 ()
fffffe00f732fcc0 mac:mac_provider_tx+85 ()
fffffe00f732fd80 mac:mac_tx+295 ()
fffffe00f732fe40 viona:viona_tx+48c ()
fffffe00f732fea0 viona:viona_worker_tx+73 ()
fffffe00f732fef0 viona:viona_worker+126 ()
fffffe00f732ff00 unix:thread_start+b ()

dumping to /dev/zvol/dsk/rpool/dump, offset 65536, content: kernel
```

I did the following:

- use qemu to boot `alpine-virt-3.17.2-x86_64.iso` and install it to an nvme drive:

```
truncate -s 2G install.raw

sudo qemu-system-x86_64 \
	  -enable-kvm \
	  -m 2g \
	  -bios /usr/share/OVMF/OVMF_CODE.fd \
	  -cdrom /home/jwm/Downloads/alpine-virt-3.17.2-x86_64.iso \
	  -drive file="install.raw",if=none,format=raw,id=nvme1 \
	  -device nvme,drive=nvme1,serial=nvme-1,addr=0x4
```

- follow the `setup-alpine` steps in https://github.com/oxidecomputer/propolis/issues/322

- halt the VM, boot it again without the ISO

```
sudo qemu-system-x86_64 \
    -enable-kvm \
    -m 2g \
    -bios /usr/share/OVMF/OVMF_CODE.fd \
    -drive file="install.raw",if=none,format=raw,id=nvme1 \
    -device nvme,drive=nvme1,serial=nvme-1,addr=0x4
```

- install packages:

```
# enable the community repo
vi /etc/apk/repositories
apk update
apk add cloud-init py3-pyserial py3-netifaces e2fsprogs e2fsprogs-extra vim git go bash build-base
```

- enable cloud init

```
setup-cloud-init
```

- set kernel serial console (alpine doesn't do this by default and I was hoping to see some Linux kernel messages during the hang)

```
vim /etc/default/grub:
	
GRUB_TERMINAL="serial console"
GRUB_CMDLINE_LINUX_DEFAULT="console=tty0 console=ttyS0,115200 modules=sd-mod,usb-storage,ext4,nvme quiet rootfstype=ext4"

grub-mkconfig -o /boot/grub/grub.cfg
```

- allow logging in on the serial console (to debug if the guest wasn't reachable by ssh):

```
vim /etc/inittab:

# Put a getty on the serial port
ttyS0::respawn:/sbin/getty -L ttyS0 115200 vt100

(comment out ttyS0 line on the bottom)
```

- disable ipv6:

```
echo 'net.ipv6.conf.all.disable_ipv6=1' >> /etc/sysctl.conf
```

- power off the VM, upload install.raw as an image (I am running Nexus with the current "import to disk" PR)

- boot an instance from this with 8 CPUs and 16G of RAM

- in the serial console, I had to set a password for the "alpine" user that cloud-init created, otherwise SSH wouldn't work

- ssh in as "alpine", then follow the repro steps in https://github.com/oxidecomputer/propolis/issues/322:

```
$ git clone http://github.com/davepacheco/go.git
$ cd go
$ git checkout go1.19.2
$ cd src
$ cat repro.sh
#!/bin/bash

# the suffix "-alpine" is used to skip some tests
export GO_BUILDER_NAME="propolis-alpine"

bash ./make.bash

while bash ./run.bash -run='!(^cgo_test$)' > test.out 2>&1;
do
    date;
done

echo "failed"
$ ./repro.sh
```

In another window, I was running `tail -F test.out`:

```
##### Building packages and commands.

##### Test execution environment.
# GOARCH: amd64
# CPU: AMD Ryzen 9 3950X 16-Core Processor            
# GOOS: linux
# OS Version: Linux 5.15.96-0-virt #1-Alpine SMP Sat, 25 Feb 2023 21:02:54 +0000 x86_64

##### Testing packages.
ok  	archive/tar	0.239s
ok  	archive/zip	0.270s
ok  	bufio	0.088s
ok  	bytes	0.695s
ok  	compress/bzip2	0.068s
ok  	compress/flate	0.644s
ok  	compress/gzip	1.072s
ok  	compress/lzw	0.083s
ok  	compress/zlib	0.353s
ok  	container/heap	0.017s
ok  	container/list	0.001s
ok  	container/ring	0.013s
ok  	context	0.040s
ok  	crypto	0.002s
ok  	crypto/aes	0.026s
ok  	crypto/cipher	0.033s
ok  	crypto/des	0.014s
ok  	crypto/dsa	0.022s
ok  	crypto/ecdsa	0.004s
ok  	crypto/ed25519	0.052s
ok  	crypto/elliptic	0.013s
ok  	crypto/hmac	0.037s
ok  	crypto/internal/boring	0.029s
ok  	crypto/internal/boring/bcache	0.207s
ok  	crypto/internal/edwards25519	4.004s
ok  	crypto/internal/edwards25519/field	2.940s
ok  	crypto/internal/nistec	0.823s
ok  	crypto/internal/nistec/fiat	0.002s [no tests to run]
ok  	crypto/internal/subtle	0.023s
ok  	crypto/md5	0.008s
ok  	crypto/rand	0.033s
ok  	crypto/rc4	0.067s
ok  	crypto/rsa	0.068s
ok  	crypto/sha1	0.029s
ok  	crypto/sha256	0.025s
ok  	crypto/sha512	0.009s
ok  	crypto/subtle	0.028s
ok  	crypto/tls	0.475s
ok  	crypto/x509	0.513s
ok  	database/sql	0.275s
ok  	database/sql/driver	0.020s
ok  	debug/buildinfo	0.017s
ok  	debug/dwarf	0.009s
ok  	debug/elf	0.019s
ok  	debug/gosym	0.307s
ok  	debug/macho	0.006s
ok  	debug/pe	0.019s
ok  	debug/plan9obj	0.005s
ok  	embed	0.030s [no tests to run]
ok  	embed/internal/embedtest	0.002s
ok  	encoding/ascii85	0.017s
ok  	encoding/asn1	0.012s
ok  	encoding/base32	0.055s
ok  	encoding/base64	0.015s
ok  	encoding/binary	0.024s
ok  	encoding/csv	0.025s
ok  	encoding/gob	0.040s
ok  	encoding/hex	0.012s
ok  	encoding/json	0.164s
ok  	encoding/pem	0.358s
ok  	encoding/xml	0.066s
ok  	errors	0.011s
ok  	expvar	0.022s
ok  	flag	0.105s
ok  	fmt	0.043s
ok  	go/ast	0.009s
ok  	go/build	1.279s
ok  	go/build/constraint	0.008s
ok  	go/constant	0.005s
ok  	go/doc	0.069s
ok  	go/doc/comment	0.742s
ok  	go/format	0.024s
ok  	go/importer	0.083s
ok  	go/internal/gccgoimporter	0.030s
ok  	go/internal/gcimporter	11.476s
ok  	go/internal/srcimporter	9.951s
ok  	go/parser	0.191s
ok  	go/printer	0.221s
ok  	go/scanner	0.019s
ok  	go/token	0.030s
ok  	go/types	5.453s
ok  	hash	0.017s
ok  	hash/adler32	0.033s
ok  	hash/crc32	0.012s
ok  	hash/crc64	0.022s
ok  	hash/fnv	0.003s
ok  	hash/maphash	0.095s
ok  	html	0.007s
ok  	html/template	0.085s
ok  	image	0.067s
ok  	image/color	0.073s
ok  	image/draw	0.061s
ok  	image/gif	10.491s
ok  	image/jpeg	2.524s
ok  	image/png	0.136s
ok  	index/suffixarray	0.125s
ok  	internal/abi	0.092s
ok  	internal/buildcfg	0.023s
ok  	internal/cpu	0.006s
ok  	internal/diff	0.009s
ok  	internal/fmtsort	0.022s
ok  	internal/fuzz	0.006s
ok  	internal/godebug	0.009s
ok  	internal/intern	0.218s
ok  	internal/itoa	0.012s
ok  	internal/poll	0.223s
ok  	internal/profile	0.008s
ok  	internal/reflectlite	0.042s
ok  	internal/singleflight	0.020s
ok  	internal/trace	0.065s
ok  	internal/unsafeheader	0.043s
ok  	internal/xcoff	0.016s
ok  	io	0.070s
ok  	io/fs	0.260s
ok  	io/ioutil	0.016s
ok  	log	0.016s
ok  	log/syslog	1.235s
ok  	math	0.010s
ok  	math/big	1.158s
ok  	math/bits	0.015s
ok  	math/cmplx	0.036s
ok  	math/rand	0.178s
ok  	mime	0.012s
ok  	mime/multipart	0.131s
ok  	mime/quotedprintable	0.066s
- exit 255
```

I saw the panic on the physical host's serial console.

**Comment by jmpesp** [comment]
```
james@gravytrain:~$ fmdump -Vp -u 17ca687c-efcb-e0af-a360-d82d0fbcd2b9
TIME                           UUID                                 SUNW-MSG-ID
Mar 03 2023 15:45:44.969376000 17ca687c-efcb-e0af-a360-d82d0fbcd2b9 SUNOS-8000-KL

  TIME                 CLASS                                 ENA
  Mar 03 15:45:44.9320 ireport.os.sunos.panic.dump_available 0x0000000000000000
  Mar 03 15:45:29.6886 ireport.os.sunos.panic.dump_pending_on_device 0x0000000000000000

nvlist version: 0
        version = 0x0
        class = list.suspect
        uuid = 17ca687c-efcb-e0af-a360-d82d0fbcd2b9
        code = SUNOS-8000-KL
        diag-time = 1677858344 939519
        de = fmd:///module/software-diagnosis
        fault-list-sz = 0x1
        fault-list = (array of embedded nvlists)
        (start fault-list[0])
        nvlist version: 0
                version = 0x0
                class = defect.sunos.kernel.panic
                certainty = 0x64
                asru = sw:///:path=/var/crash/gravytrain/.17ca687c-efcb-e0af-a360-d82d0fbcd2b9
                resource = sw:///:path=/var/crash/gravytrain/.17ca687c-efcb-e0af-a360-d82d0fbcd2b9
                savecore-succcess = 1
                dump-dir = /var/crash/gravytrain
                dump-files = vmdump.3
                os-instance-uuid = 17ca687c-efcb-e0af-a360-d82d0fbcd2b9
                panicstr = panicked at 'called `Result::unwrap()` on an `Err` value: MaxCapacity(8096)', /work/oxidecomputer/opte/opte/src/engine/port.rs:1890:17
                panicstack = xde:rust_begin_unwind+146 () | xde:_ZN4core9panicking9panic_fmt17hbd42276bcaa2dfd8E+2d () | xde:_ZN4core6result13unwrap_failed17h5b67be868f6dc91bE+6c () | xde:_ZN4opte6engine4port13Port$LT$N$GT$7process17h9bca109443cf9deaE+2bdf () | xde:xde_mc_tx+32f () | mac:mac_ring_tx+43 () | mac:mac_provider_tx+85 () | mac:mac_tx+295 () | viona:viona_tx+48c () | viona:viona_worker_tx+73 () | viona:viona_worker+126 () | unix:thread_start+b () | 
                crashtime = 1677858238
                panic-time = Fri Mar  3 15:43:58 2023 UTC
        (end fault-list[0])

        fault-status = 0x1
        severity = Major
        __ttl = 0x1
        __tod = 0x64021628 0x39c78100


```

**Comment by jmpesp** [comment]
Digging into the crash dump:

```
# mdb vmcore.3
Loading modules: [ unix genunix specfs dtrace mac cpu.generic uppc apix scsi_vhci zfs sata ip hook neti sockfs arp usba xhci sd smbios stmf stmf_sbd mm lofs random crypto cpc ufs logindmux nsmb ptm smbsrv nfs ]
> $G
C++ symbol demangling enabled
> ::status
debugging crash dump vmcore.3 (64-bit) from gravytrain
operating system: 5.11 helios-1.0.21472 (i86pc)
build version: heads/master-0-g8b0ccc49c4

image uuid: 17ca687c-efcb-e0af-a360-d82d0fbcd2b9
panic message: panicked at 'called `Result::unwrap()` on an `Err` value: MaxCapacity(8096)', /work/oxidecomputer/opte/opte/src/engine/port.rs:1890:17
dump content: kernel pages only
> $C
fffffe00f732edb0 vpanic()
fffffe00f732ee40 rust_begin_unwind+0x146()
fffffe00f732ee80 core::panicking::panic_fmt::hbd42276bcaa2dfd8+0x2d()
fffffe00f732ef00 core::result::unwrap_failed::h5b67be868f6dc91b+0x6c()
fffffe00f732f310 opte::engine::port::Port<N>::process::h9bca109443cf9dea+0x2bdf()
fffffe00f732fc40 xde_mc_tx+0x32f()
fffffe00f732fc70 mac_ring_tx+0x43(fffffeb1ddf41010, 0, fffffeb1f8e2c460)
fffffe00f732fcc0 mac_provider_tx+0x85(fffffeb1ddf41010, 0, fffffeb1f8e2c460, fffffeb2761e0e40)
fffffe00f732fd80 mac_tx+0x295(fffffeb2761e0e40, fffffeb1f8e2c460, 0, 1, 0)
fffffe00f732fe40 viona_tx+0x48c(fffffeb2747eacc0, fffffeb2747eae00)
fffffe00f732fea0 viona_worker_tx+0x73(fffffeb2747eae00, fffffeb2747eacc0)
fffffe00f732fef0 viona_worker+0x126(fffffeb2747eae00)
fffffe00f732ff00 thread_start+0xb()
> 0xfffffeb1f8e2c460::walk b_cont mblk_t | ::mblk
            ADDR FL TYPE    LEN   BLEN              RPTR             DBLK
fffffeb1f8e2c460 0  data    81    138   fffffeb2b90446f8 fffffeb1f8e25f00
> 0xfffffeb1f8e2c460::print mblk_t
{
    b_next = 0
    b_prev = 0
    b_cont = 0
    b_rptr = 0xfffffeb2b90446f8
    b_wptr = 0xfffffeb2b9044749
    b_datap = 0xfffffeb1f8e25f00
    b_band = 0
    b_tag = 0
    b_flag = 0
    b_queue = 0
}
```

**Comment by jmpesp** [comment]
```
> 0xfffffeb2b90446f8::dump -u -l 0x51 -t -g 1     
                    8  9  a  b  c  d  e  f   0  1  2  3  4  5  6  7  89abcdef01234567
fffffeb2b90446f8:  a8 40 25 ff 77 77 a8 40  25 fa 4c ce 08 00 45 00  .@%.ww.@%.L...E.
fffffeb2b9044708:  00 43 ee d6 40 00 40 11  8f a0 ac 1e 00 05 08 08  .C..@.@.........
fffffeb2b9044718:  08 08 e3 60 00 35 00 2f  6c 9e e4 67 01 00 00 01  ...`.5./l..g....
fffffeb2b9044728:  00 00 00 00 00 01 06 67  6f 6c 61 6e 67 03 6f 72  .......golang.or
fffffeb2b9044738:  67 00 00 01 00 01 00 00  29 04 d0 00 00 00 00 00  g.......).......
fffffeb2b9044748:  00
```

Reassembling this packet into a format wireshark can import:

```
I 2023-03-03T16:08:46Z
00000000: a8 40 25 ff 77 77 a8 40 25 fa 4c ce 08 00 45 00  .@%.ww.@%.L...E.
00000010: 00 43 ee d6 40 00 40 11 8f a0 ac 1e 00 05 08 08  .C..@.@.........
00000020: 08 08 e3 60 00 35 00 2f 6c 9e e4 67 01 00 00 01  ...`.5./l..g....
00000030: 00 00 00 00 00 01 06 67 6f 6c 61 6e 67 03 6f 72  .......golang.or
00000040: 67 00 00 01 00 01 00 00 29 04 d0 00 00 00 00 00  g.......).......
00000050: 00                                               .
```

It looks like a DNS query?

```
Frame 1: 81 bytes on wire (648 bits), 81 bytes captured (648 bits) on interface unknown, id 0
    Interface id: 0 (unknown)
        Interface name: unknown
    Encapsulation type: Ethernet (1)
    Arrival Time: Mar  3, 2023 11:13:11.000001000 EST
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1677859991.000001000 seconds
    [Time delta from previous captured frame: 0.000000000 seconds]
    [Time delta from previous displayed frame: 0.000000000 seconds]
    [Time since reference or first frame: 0.000000000 seconds]
    Frame Number: 1
    Frame Length: 81 bytes (648 bits)
    Capture Length: 81 bytes (648 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: eth:ethertype:ip:udp:dns]
    [Coloring Rule Name: UDP]
    [Coloring Rule String: udp]
Ethernet II, Src: OxideCom_fa:4c:ce (a8:40:25:fa:4c:ce), Dst: OxideCom_ff:77:77 (a8:40:25:ff:77:77)
    Destination: OxideCom_ff:77:77 (a8:40:25:ff:77:77)
        Address: OxideCom_ff:77:77 (a8:40:25:ff:77:77)
        .... ..0. .... .... .... .... = LG bit: Globally unique address (factory default)
        .... ...0 .... .... .... .... = IG bit: Individual address (unicast)
    Source: OxideCom_fa:4c:ce (a8:40:25:fa:4c:ce)
        Address: OxideCom_fa:4c:ce (a8:40:25:fa:4c:ce)
        .... ..0. .... .... .... .... = LG bit: Globally unique address (factory default)
        .... ...0 .... .... .... .... = IG bit: Individual address (unicast)
    Type: IPv4 (0x0800)
Internet Protocol Version 4, Src: 172.30.0.5, Dst: 8.8.8.8
    0100 .... = Version: 4
    .... 0101 = Header Length: 20 bytes (5)
    Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)
        0000 00.. = Differentiated Services Codepoint: Default (0)
        .... ..00 = Explicit Congestion Notification: Not ECN-Capable Transport (0)
    Total Length: 67
    Identification: 0xeed6 (61142)
    Flags: 0x40, Don't fragment
        0... .... = Reserved bit: Not set
        .1.. .... = Don't fragment: Set
        ..0. .... = More fragments: Not set
    ...0 0000 0000 0000 = Fragment Offset: 0
    Time to Live: 64
    Protocol: UDP (17)
    Header Checksum: 0x8fa0 [validation disabled]
    [Header checksum status: Unverified]
    Source Address: 172.30.0.5
    Destination Address: 8.8.8.8
User Datagram Protocol, Src Port: 58208, Dst Port: 53
    Source Port: 58208
    Destination Port: 53
    Length: 47
    Checksum: 0x6c9e [unverified]
    [Checksum Status: Unverified]
    [Stream index: 0]
    [Timestamps]
        [Time since first frame: 0.000000000 seconds]
        [Time since previous frame: 0.000000000 seconds]
    UDP payload (39 bytes)
Domain Name System (query)
    Transaction ID: 0xe467
    Flags: 0x0100 Standard query
        0... .... .... .... = Response: Message is a query
        .000 0... .... .... = Opcode: Standard query (0)
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...1 .... .... = Recursion desired: Do query recursively
        .... .... .0.. .... = Z: reserved (0)
        .... .... ...0 .... = Non-authenticated data: Unacceptable
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 1
    Queries
        golang.org: type A, class IN
            Name: golang.org
            [Name Length: 10]
            [Label Count: 2]
            Type: A (Host Address) (1)
            Class: IN (0x0001)
    Additional records
        <Root>: type OPT
            Name: <Root>
            Type: OPT (41)
            UDP payload size: 1232
            Higher bits in extended RCODE: 0x00
            EDNS0 version: 0
            Z: 0x0000
                0... .... .... .... = DO bit: Cannot handle DNSSEC security RRs
                .000 0000 0000 0000 = Reserved: 0x0000
            Data length: 0
```

**Comment by bnaecker** [comment]
The packet itself doesn’t seem important. This is panicking when trying to add a new entry to the flow table. We statically limit that to 8096 TCP flows, and there’s an unwrap trying to add one when we’re out of space.

**Comment by jmpesp** [comment]
[346.pcap.zip](https://github.com/oxidecomputer/opte/files/10884035/346.pcap.zip)

**Comment by jmpesp** [comment]
Relevant limits:

https://github.com/oxidecomputer/opte/blob/b274cad9e981cbb7de1d3a48834b1f97500a13a8/xde/src/xde.rs#L96-L106

These seem a little low, no? Shouldn't they be 65536?

**Comment by bnaecker** [comment]
I think it depends on how the control plane configures the guest. But it doesn’t seem right, since the guest should have at least 16k ports, or up to 65k if this has an ephemeral IP.

**Comment by rmustacc** [comment]
Regardless of the limit, we can't panic unwrap when we hit it.

**Comment by bnaecker** [comment]
Agreed. There's a TODO on it and it's part of the issue around auditing panics.

**Comment by bnaecker** [comment]
I'm going to improve this by limiting the flow tables to the number of ports we allocate the the guests. This is in the `VpcCfg`. I'm not sure how this will shake out, since folks can have dual stack, which would double the flow-table sizes. That shouldn't matter until we actually support IPv6 guests, so it seems okay to me. It's also not clear the distinction between UFT flows and TCP flows. I think the former is a superset of the latter, but I've not been in this code for a while.

**Comment by luqmana** [comment]
> It's also not clear the distinction between UFT flows and TCP flows. I think the former is a superset of the latter, but I've not been in this code for a while.

Sort of, the TCP flow table is just for cleaning up entries in the main Unified Flow Table (UFT) on TCP close/reset.

https://github.com/oxidecomputer/opte/blob/b274cad9e981cbb7de1d3a48834b1f97500a13a8/opte/src/engine/port.rs#L615-L642

**Comment by bnaecker** [comment]
Closed by #347. But see #358 which tracks a better solution to the problem.
